### PR TITLE
Feat: files processing

### PR DESCRIPTION
**Pull request by amirRamirfatahi**
# Pre-submission Checklist

> For tests to work you need a working neo4j and redis instance with the example dataset in `docker/db-graph`

- [ ] **Testing**: Implement and pass new tests for the new features/fixes, `cargo test`.
- [ ] **Performance**: Ensure new code has relevant performance benchmarks, `cargo bench`

### CODE REVIEW

File: src/static_processor/store.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    let path_exists = match fs::metadata(path.as_str()).await {
+        Err(_) => false,
+        Ok(metadata) => metadata.is_dir(),
+    };
+
+    if !path_exists {
+        fs::create_dir_all(path.as_str()).await?;
+    }
+
+    let file_path = format!("{}/{}", path, name);
```
Comment (SHAcollision) [inline src/static_processor/store.rs:19]: Better use `PathBuf` and .join() to build paths reliably instead all those `format!("{}/{}/{}..."` ?

File: src/static_processor/mod.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        details::{FileUrls, FileVersions},
+        FileDetails,
+    },
+    types::DynError,
+};
+
+pub mod image;
+pub mod store;
+pub mod video;
+
```
Comment (SHAcollision) [inline src/static_processor/mod.rs:12]: `src/static_processor` is a new top level module with a bunch of miscellaneous utils. So much room to use Rust idiomatic for better maintainability and readability. 

ideas
```
pub trait MediaProcessor {
    fn create_variant(&self, version: FileVersion) -> Result<String, DynError>;
}

impl MediaProcessor for ImageDetails {
    fn create_variant(&self, version: FileVersion) -> Result<String, DynError> {
        // Implement image-specific logic
    }
}

impl MediaProcessor for VideoDetails {
    fn create_variant(&self, version: FileVersion) -> Result<String, DynError> {
        // Implement video-specific logic
    }
}
```

Please, document the new functions.

File: src/events/handlers/file.rs [hunk 3]
```diff
@@ -64,48 +60,16 @@ async fn ingest(
...
-        urls: FileUrls { main: static_path },
-    })
-}
-
-async fn store_blob(name: String, path: String, blob: &Bytes) -> Result<(), DynError> {
-    let storage_path = Config::from_env().file_path;
+    let path: String = format!("{}/{}", user_id, file_id);
+    let storage_path = get_storage_path();
     let full_path = format!("{}/{}", storage_path, path);
+    store_blob(String::from("main"), full_path.to_string(), &blob).await?;
```
Comment (SHAcollision) [inline src/events/handlers/file.rs:69]: Is `"main"` the "name" ? or the variant?, fix `store_blob()` signature.

File: src/static_processor/video.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+//     } else {
+//         Err(IoError::new(
+//             ErrorKind::Other,
+//             format!(
+//                 "FFmpeg metadata extraction failed: {}",
+//                 String::from_utf8_lossy(&child_output.stderr)
+//             ),
+//         ))
+//     }
+// }
```
Comment (SHAcollision) [inline src/static_processor/video.rs:117]: ?

File: src/static_processor/video.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+pub fn get_video_versions() -> Vec<FileVersions> {
+    vec![]
+}
```
Comment (SHAcollision) [inline src/static_processor/video.rs:121]: ?

File: src/static_processor/mod.rs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    };
+
+    FileUrls {
+        main: format!("{}/main", path),
+        feed: versions
+            .contains(&FileVersions::FEED)
+            .then_some(format!("{}/feed", path)),
+        small: versions
+            .contains(&FileVersions::SMALL)
+            .then_some(format!("{}/small", path)),
```
Comment (SHAcollision) [inline src/static_processor/mod.rs:54]: `".../small"`

You implemented the Display trait on `FileVersions` enum.

File: src/routes/static.rs [hunk 7]
```diff
@@ -1,50 +1,140 @@
...
-                    );
-                    response
-                        .headers_mut()
-                        .insert("content-type", value.content_type.parse().unwrap());
-                    Ok(response)
-                }
-                None => Err(StatusCode::NOT_FOUND),
+    // for backward compatibility with old urls, replace old urls with main version of file
+    let new_uri = if path_parts.len() == 5 {
+        format!("{}/main", request.uri().path())
```
Comment (SHAcollision) [inline src/routes/static.rs:51]: you implemented Display on the FileVersions enum. Could be used here.

File: src/routes/static.rs [hunk 8]
```diff
@@ -1,50 +1,140 @@
...
+
+    if version.is_none() {
+        return Err(StatusCode::NOT_FOUND);
+    }
+
+    let files =
+        match FileDetails::get_by_ids(vec![vec![user_id, file_id].as_slice()].as_slice()).await {
+            Ok(files) => files,
+            Err(_) => {
+                return Err(StatusCode::INTERNAL_SERVER_ERROR);
```
Comment (SHAcollision) [inline src/routes/static.rs:76]: Is this the best descriptive error we can throw here?

File: src/static_processor/image.rs [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    let origin_path = format!("{}/{}/{}", get_storage_path(), file.owner_id, file.id,);
+
+    let origin_file = format!("{}/main", origin_path,);
+
+    let options = image_options.unwrap();
+
+    let image_ext = get_image_format(origin_file.as_str()).await.unwrap();
+
+    let output_format = match image_ext.as_str() {
+        "JPEG" => None,
```
Comment (SHAcollision) [inline src/static_processor/image.rs:60]: If we are going to re-format, we could very well better use some WEBP compressed to ~80% quality

**Review by SHAcollision — CHANGES_REQUESTED** [review]
TLDR: this PR needs your own detailed review, there is a lot of low hanging improvements that hint you have not gone over your PR before requesting the review: this leads to a waste of time for both.

Certainly needs:
- Rewrite `static_files_middleware` to make it readable. If possible, let's use Axum handlers and document this endpoint on swagger as well.
- Benchmarking is needed.
- Provide Clear Documentation & Comments
If `static_processor` has a variety of specialized logic (e.g., ImageMagick/FFmpeg commands, path manipulations, version enumerations, caching, etc.), make sure to:
1. Document each function with a short doc comment (///).
2. Explain why certain commands or transformations are used, especially if they involve system calls or external tools (ImageMagick, FFmpeg).
3. Clarify any assumptions about the environment (e.g., “We expect convert from ImageMagick to be installed.”). And of course: make it optional!
- Although tests exist, some scenarios require more robust checks:
Ensure tests handle invalid file versions (e.g., a request for a “nonexistent” variant).
Validate edge cases like large file sizes, unsupported content types, or corrupted data.

Overall a lot of simplification, better error handling, improved readability/maintainability, and stick to the conventions we use in the rest of the codebase is needed.

File: src/static_processor/video.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        &input_path,
+        version,
+        options.width.as_str(),
+        Some(options.format.as_str()),
+    )
+    .await
+    {
+        Ok(_) => Ok(options.content_type),
+        Err(err) => Err(format!("Failed to process video: {}", err).into()),
+    }
```
Comment (SHAcollision) [inline src/static_processor/video.rs:59]: Pretty weird construction here again. A `match` statement on the return of a function that calls an external process to later do a half-assed handling of the result. Can't you simply make sure `process_video` returns the correct error and raise with `process_video().await?`

File: src/routes/static.rs [hunk 11]
```diff
@@ -1,50 +1,140 @@
+use std::borrow::BorrowMut;
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static.rs]: what do you mean "??" :)) 
Had to import it because of `*request.uri_mut() = new_uri.parse().unwrap();`.  Both are removed.

File: src/routes/static.rs [hunk 12]
```diff
@@ -1,50 +1,140 @@
...
-                None => Err(StatusCode::NOT_FOUND),
+    // for backward compatibility with old urls, replace old urls with main version of file
+    let new_uri = if path_parts.len() == 5 {
+        format!("{}/main", request.uri().path())
+    } else if path_parts[5].is_empty() {
+        format!("{}main", request.uri().path())
+    } else {
+        request.uri().path().to_string()
+    };
+    *request.borrow_mut().uri_mut() = new_uri.parse().unwrap();
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static.rs]: yes..

File: src/routes/static.rs [hunk 13]
```diff
@@ -1,50 +1,140 @@
...
+async fn static_files_middleware(
+    params: Query<FileParams>,
+    mut request: Request,
+    next: Next,
+) -> Result<Response, StatusCode> {
     let path = String::from(request.uri().path());
 
-    let mut response = next.run(request).await;
+    // path_parts: ["", "static", "files", "<USER_ID>", "<FILE_ID>", "<VERSION_NAME>"]
+    let path_parts: Vec<&str> = path.split("/").collect();
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static.rs]: Will try the extractor now. previously couldn't because it didn't support it.

File: src/routes/static.rs [hunk 14]
```diff
@@ -1,50 +1,140 @@
...
+    let new_uri = if path_parts.len() == 5 {
+        format!("{}/main", request.uri().path())
+    } else if path_parts[5].is_empty() {
+        format!("{}main", request.uri().path())
+    } else {
+        request.uri().path().to_string()
+    };
+    *request.borrow_mut().uri_mut() = new_uri.parse().unwrap();
+
+    let version = if path_parts.len() == 6 {
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static.rs]: this is a custom middleware for this one handler (possibly for a long time). I don't see how it can be a nightmare?!

File: src/routes/static.rs [hunk 15]
```diff
@@ -1,50 +1,140 @@
...
+
+    if fs::metadata(path).await.is_ok() {
+        return Ok(None);
+    }
+
+    match static_processor::create_file_version(file, version).await {
+        Ok(content_type) => Ok(Some(content_type)),
+        Err(err) => {
+            debug!("Failed to create file version: {:?}", err);
+            Err(StatusCode::INTERNAL_SERVER_ERROR)
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static.rs]: What do you mean "wasteful error management?!"

File: src/static_processor/image.rs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+fn get_image_option_for_version(
+    file: &FileDetails,
+    version: &FileVersions,
+) -> Option<ImageOptions> {
+    let format = match file.content_type.as_str() {
+        "image/gif" => String::from("gif"),
+        _ => String::from("jpeg"),
+    };
+    let width = match version {
+        FileVersions::SMALL => String::from("320"),
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/static_processor/image.rs]: resolution will be kept the same. that's why we only need a width.

File: src/static_processor/image.rs (hunk 9 above)
Comment (amirRamirfatahi) [reply src/static_processor/image.rs]: It still doesn't have the availability of jpeg 
https://en.wikipedia.org/wiki/Comparison_of_web_browsers#Image_format_support

Comment (SHAcollision) [reply src/static_processor/image.rs]: All of the assets of pubky-app frontend are WEBP. It's rare to find browsers that do not support WEBP. We should default to webp and only if there were issues (I guess not) do JPEG.

File: src/static_processor/middleware.rs [hunk 17]
```diff
@@ -0,0 +1,167 @@
+use axum::{
+    extract::{Query, Request},
+    middleware::Next,
+    response::{IntoResponse, Redirect, Response},
+};
+use log::error;
+use reqwest::StatusCode;
```
Comment (SHAcollision) [inline src/static_processor/middleware.rs:7]: We are building an **Axum** server, so we should stick to using Axum types.

Using `reqwest` client `StatusCode` type for our Axum response is sketchy. In addition, as soon as PubkyClient can perform requests to the `{homeserver_pubky_id}/events` we won't need `Reqwest` library anymore.

File: src/static_processor/processors/image.rs [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+            .arg(output)
+            .output() // Automatically pipes stdout and stderr
+            .await?;
+
+        if child_output.status.success() {
+            Ok(String::from_utf8_lossy(&child_output.stdout).to_string())
+        } else {
+            Err(format!(
+                "ImageMagick command failed: {}",
+                String::from_utf8_lossy(&child_output.stdout)
```
Comment (tipogi) [inline src/static_processor/processors/image.rs:98]: Maybe we should use `stderr` for more accurate error handling?

File: src/static_processor/processors/image.rs [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+            false => format!("{}:{}", options.format, output_file_path),
+        };
+
+        let child_output = Command::new("convert")
+            .arg(origin_file_path)
+            .arg("-background")
+            .arg("white")
+            .arg("-alpha")
+            .arg("remove")
+            .arg("-resize")
```
[the remark this answers is not in the document]
Comment (tipogi) [reply src/static_processor/processors/image.rs:87]: Also maybe add `-strip` command to delete metadata?

File: src/static_processor/store.rs [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+        if !path_exists {
+            fs::create_dir_all(path.as_str()).await?;
+        }
+
+        let file_path = format!("{}/{}", path, name);
+        let mut static_file = File::create_new(file_path).await?;
+        static_file.write_all(blob).await?;
+
+        Ok(())
```
Comment (tipogi) [inline src/static_processor/store.rs:26]: refactor

File: src/events/handlers/file.rs
Comment (tipogi) [inline src/events/handlers/file.rs:1]: In the `put` function, we usually convert the event to `pubky-app-specs` mdoel and then send that object to `sync_put`. We no longer use `ingest` name

File: src/routes/static/files.rs [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+    };
+
+    if files.is_empty() {
+        return Err(Error::FileNotFound {});
+    }
+
+    let file = match files[0].clone() {
+        Some(file) => file,
+        None => return Err(Error::FileNotFound {}),
+    };
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static/files.rs:71]: I'm returning from the function which you can't do in a closure.  
The rest is also much more readable IMO.

File: src/routes/static/files.rs [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+                "variant {} is not valid for content type {}",
+                variant, file.content_type
+            ),
+        });
+    }
+
+    let file_variant_exists = StaticProcessor::check_variant_exists(&file, variant.clone()).await;
+
+    let file_variant_content_type = if file_variant_exists {
+        StaticProcessor::get_content_type_for_variant(&file, &variant)
```
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/routes/static/files.rs:88]: that was verification that a variant for a content_type exists.  
This is to get the new processed file content type.

File: src/static_processor/processors/image.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (amirRamirfatahi) [reply src/static_processor/processors/image.rs:87]: I think this way is really cleaner.  
Not sure about deleting metadata though.

File: src/static_processor/processors/image.rs (hunk 18 above)
Comment (amirRamirfatahi) [reply src/static_processor/processors/image.rs:98]: `.output() // Automatically pipes stdout and stderr` is a couple lines up

File: src/events/handlers/file.rs
Comment (amirRamirfatahi) [reply src/events/handlers/file.rs:1]: this is already happening? `ingest` is another function.

File: src/static_processor/processors/image.rs [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+            Ok(String::from_utf8_lossy(&child_output.stdout).to_string())
+        } else {
+            Err(format!(
+                "ImageMagick format extraction failed: {}",
+                String::from_utf8_lossy(&child_output.stderr)
+            )
+            .into())
+        }
+    }
+}
```
Comment (SHAcollision) [inline src/static_processor/processors/image.rs:129]: Suggestion for simplification and use of `cwebp`
```suggestion
const SMALL_IMAGE_WIDTH: &str = "320";
const FEED_IMAGE_WIDTH: &str = "720";

pub struct ImageOptions {
    width: Option<String>,
    content_type: String,
}

impl BaseProcessingOptions for ImageOptions {
    fn content_type(&self) -> String {
        self.content_type.clone()
    }
}

pub struct ImageProcessor;

#[async_trait]
impl FileProcessor for ImageProcessor {
    type ProcessingOptions = ImageOptions;

    fn get_valid_variants_for_content_type(_content_type: &str) -> Vec<FileVariant> {
        // Always allow main, feed, small for images including gif.
        vec![FileVariant::Main, FileVariant::Small, FileVariant::Feed]
    }

    fn get_content_type_for_variant(_file: &FileDetails, _variant: &FileVariant) -> String {
        // We always convert to image/webp
        String::from("image/webp")
    }

    fn get_options_for_variant(
        file: &FileDetails,
        variant: &FileVariant,
    ) -> Result<ImageOptions, DynError> {
        // We always produce .webp
        let content_type = Self::get_content_type_for_variant(file, variant);
        let width = match variant {
            // For 'main', we won't resize. For feed/small, we specify widths.
            FileVariant::Main => None,
            FileVariant::Feed => Some(String::from(FEED_IMAGE_WIDTH)),
            FileVariant::Small => Some(String::from(SMALL_IMAGE_WIDTH)),
        };
        Ok(ImageOptions {
            width,
            content_type,
        })
    }

    async fn process(
        origin_file_path: &str,
        output_file_path: &str,
        options: &ImageOptions,
    ) -> Result<String, DynError> {
        // We will call cwebp to convert to .webp at 80% quality
        // If a width is specified, we do:  -resize <width> 0
        // Otherwise, no -resize argument => keeps original size
        //
        // We'll produce an output file without .webp in name to stay consistent
        // with how code references "main", "feed", "small". But the content is .webp.

        use tokio::process::Command;

        let mut cmd_args = vec!["-q", "80"];
        if let Some(ref w) = options.width {
            cmd_args.push("-resize");
            cmd_args.push(w);
            cmd_args.push("0");
        }
        cmd_args.push(origin_file_path);
        cmd_args.push("-o");
        cmd_args.push(output_file_path);

        let child_output = Command::new("cwebp").args(&cmd_args).output().await?;

        if !child_output.status.success() {
            let stderr = String::from_utf8_lossy(&child_output.stderr);
            return Err(format!("cwebp command failed: {}", stderr).into());
        }

        // Return final content_type
        Ok(options.content_type())
    }
}
```

File: .github/workflows/test.yml [hunk 24]
```diff
@@ -57,8 +57,8 @@ jobs:
       - name: Create example graph
         run: docker exec neo4j bash /db-graph/run-queries.sh
 
-      # - name: Install dependencies
-      #   run: cargo build --release
+      - name: Install dependencies
+        run: sudo apt install -y imagemagick
```
Comment (SHAcollision) [inline .github/workflows/test.yml:61]:
```suggestion
        run: sudo apt-get install -y webp
```

File: src/static_processor/processors/image.rs [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+        file: &FileDetails,
+        variant: &FileVariant,
+    ) -> Result<ImageOptions, DynError> {
+        let format = match file.content_type.as_str() {
+            "image/gif" => String::from("gif"),
+            _ => String::from("jpeg"),
+        };
+        let width = match variant {
+            FileVariant::Small => String::from(SMALL_IMAGE_WIDTH),
+            FileVariant::Feed => String::from(FEED_IMAGE_WIDTH),
```
Comment (SHAcollision) [inline src/static_processor/processors/image.rs:56]: Unresolve comment: what happens if a user uploads a 200x200 image, are we resizing it to 720?

**Review by SHAcollision — CHANGES_REQUESTED** [review]
I'm eager to merge this so we can tackle https://github.com/pubky/pubky-nexus/issues/319 but this PR is still very green IMO. @amirRamirfatahi why not focus on finishing pending PRs instead of opening more fronts (e.g. #314).

File: src/routes/v0/user/avatar.rs
[the remark this answers is not in the document]
Comment (SHAcollision) [reply src/routes/v0/user/avatar.rs:1]: Yes, I think we should move the logic inside the avatar handler to a `UserAvatar::get_by_id()` . This will also be the function we can add to benchmarks.

File: src/events/handlers/file.rs [hunk 26]
```diff
@@ -70,18 +63,23 @@ async fn ingest(
...
     match pubky_app_object {
         PubkyAppObject::Blob(blob) => {
-            store_blob(file_id.to_string(), user_id.to_string(), &blob.0).await?;
-
-            Ok(FileMeta {
-                urls: FileUrls {
-                    main: format!("{}/{}", user_id, file_id),
-                },
-            })
+            StaticStorage::store_blob(FileVariant::Main.to_string(), full_path.to_string(), &blob)
```
Comment (SHAcollision) [inline src/events/handlers/file.rs:75]: Nice

File: src/routes/v0/user/avatar.rs [hunk 27]
```diff
@@ -63,33 +66,65 @@ pub async fn user_avatar_handler(Path(user_id): Path<String>) -> Result<Response
...
+            "Invalid content type header: {}",
+            small_variant_content_type
+        );
+        Error::InternalServerError {
+            source: Box::new(err),
+        }
+    })?;
+    response
+        .headers_mut()
+        .insert("content-type", content_type_header);
```
Comment (tipogi) [inline src/routes/v0/user/avatar.rs:127]: I think that one, we could refactor creating another static function in `StaticProcessor` that returns a response as a result. we have the same block in the `static_files_handler`. One option
```rust
pub async fn swap_serve_dir_and_add_content_type(
        request: Request<Body>,
        content_type: String,
        custom_uri: Option<String>
    ) -> Result<Response<ServeFileSystemResponseBody>, Error> {
        let (request_parts, request_body) = request.into_parts();
        let mut req = Request::from_parts(request_parts.clone(), request_body);
        let file_uri_path = match custom_uri {
            Some(new_uri) => new_uri,
            None => request_parts.uri.to_string()
        };
        *req.uri_mut() =
            file_uri_path
                .as_str()
                .parse::<Uri>()
                .map_err(|err| Error::InternalServerError {
                    source: Box::new(err),
                })?;
        let response_result = get_serve_dir().try_call(req).await;

        let mut response = match response_result {
            Ok(response) => {
                if response.status() != StatusCode::OK {
                    return Ok(response);
                }
                response
            }
            Err(err) => {
                return Err(Error::InternalServerError {
                    source: Box::new(err),
                });
            }
        };

        // set the content type header
        let content_type_header = content_type.parse().map_err(|err| {
            error!("Invalid content type header: {}", content_type);
            Error::InternalServerError {
                source: Box::new(err),
            }
        })?;
        response
            .headers_mut()
            .insert("content-type", content_type_header);
        Ok(response)
    }
```
    
That refactoring makes me think whether we should keep _SERVE_DIR_INSTANCE_ in `mod.rs`, or if we should create another file inside the static folder for all `ServeDir` related operations.

Like this, we could have:

```rust
struct xxx {
fn serve_static_dir_response() -> Res<Resp,xxx> // not 100% sure the name...
fn add_content_type(&response)
fn add_content_disposition(&response)
...
static SERVE_DIR_INSTANCE...
```

**Review by tipogi — APPROVED** [review]
From my side, clear and ready to merge 🚀

**Review by SHAcollision — APPROVED** [review]
Did not test myself, but looks good. Let's figure out next steps after merge.

**Comment by SHAcollision** [comment]
@amirRamirfatahi  this PR gives me nightmares at night. Please merge.

File: src/routes/static/mod.rs [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+    let config = Config::from_env();
+
+    let router = register_routes!(
+        Router::new(),
+        STATIC_FILES_ROUTE => files::static_files_handler,
+        LEGACY_STATIC_FILES_ROUTE => legacy_files::legacy_files_handler,
+        USER_AVATAR_ROUTE => avatar::user_avatar_handler,
+    );
+
+    router.nest_service(STATIC_ROUTE, get_service(ServeDir::new(config.static_path)))
```
Comment (tipogi) [inline src/routes/static/mod.rs:25]: could we use here, `PubkyServeDir::get_serve_dir()` instead of create a new _ServeDir_?

**Comment by SHAcollision** [comment]
This PR was merged to `dev` and then `main`
